sound/soc/sof/loader.c

```diff
@@ -109,6 +109,8 @@ EXPORT_SYMBOL(snd_sof_load_firmware_memcpy);
 int snd_sof_run_firmware(struct snd_sof_dev *sdev)
 {
 	int ret;
+	struct sof_ipc_cmd_hdr hdr;
+	struct sof_ipc_reply reply;
```

Comment:
this can be moved inside of the if() case

Reply:
> In this setup the firmware will never going to send the FW_READY notification or it is going to be randomly sent too early, thus the host is not going to receive it?

FW_READY notification does get sent too early (platform driver is not even inserted yet) but I'm not sure if it gets lost since I haven't tested this case (which in hindsight seems like a poor choice on my part).

> It might be simpler to handle this use case buy using different description of the setup:
bool query_fw_ready or no_fw_ready_notification flag for such platforms.
if it is set, skip waiting for the fw_ready and head directly to ""firmware boot complete" state
implement the sof_ipc_ops.post_fw_boot for ipc3 and if the flag is set then query the fw_ready.
This way most of the changes will happen in ipc3.c and probably better layered?

IMO `init_fw_ready` should be just as clear as `query_fw_ready`. The host basically initialises the FW_READY sequence if `init_fw_ready == true`. I wouldn't mind changing its name if you really think `query_fw_ready` adds much more clarity.

Hm, using `post_fw_boot` would move all the logic from loader to IPC3 but logically speaking it doesn't make much sense IMO. The FW is assumed to have booted successfully yet you still need to query its state in `post_fw_boot`?

Reply:
@LaurentiuM1234, thanks for the pointer to the diagrams, it is clear now.

In this setup you are not going to reboot Zephyr/SOF unless you do a full system suspend to turn off all A55s, right?
On resume the INMATE will boot earlier as well due to jailhouse.

The first and biggest issue with the current PR is that you introduce IPC3 message sending in generic, IPC neutral code (loader.c), this must be avoided.

Before making  any change in architecture, I would check the following:
- The firmware sends FW_READY on boot (unless you have not disabled it explicitly)
- The message is placed in the mailbox, irq triggered (or doorbell rung) but the ROOT CELL did not handled it
- When the Linux stack boots you still have the FW_READY message in mailbox
- In theory you should be able to read it out by 'faking' an interrupt + ipc rx of a FW_READY

Can you test this?
I would use the snd_sof_dsp_run() platform code via delayed work (or just a direct call?) to craft a message and pretend that you have received the FW_READY.
I would likely check if the FW has booted up (check the mailbox for the FW_READY?), but that is just me over cautious.

If this is working then you don't need changes in core at all as the platform's peculiarities will be handled in platform code.

Reply:
Sorry for taking so long to reply. See answers below:

> The first and biggest issue with the current PR is that you introduce IPC3 message sending in generic, IPC neutral code (loader.c), this must be avoided.

This can easily be solved by moving the IPC3 code from loader.c to `imx93_run`.

> The firmware sends FW_READY on boot (unless you have disabled it explicitly)
The message is placed in the mailbox, irq triggered (or doorbell rung) but the ROOT CELL did not handled it
When the Linux stack boots you still have the FW_READY message in mailbox
In theory you should be able to read it out by 'faking' an interrupt + ipc rx of a FW_READY

Just did a small test in which the firmware sends FW_READY and `init_fw_ready=false` which resulted in a kernel panic. Seems like the IRQ came during the dsp-ipc driver probe.

> In theory you should be able to read it out by 'faking' an interrupt + ipc rx of a FW_READY

What do you mean fake an interrupt? Also, wouldn't it be more natural to just have the host initiate the FW_READY sequence (of course, all the IPC3 code would me moved from `loader.c` to the imx93 platform driver). Also, support for this flow has been added to the firmware.